packages/contracts/contracts/ProxyAdmin.sol

```diff
@@ -1,28 +1,172 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.9;
 
+import { ChugSplashRegistry } from "./ChugSplashRegistry.sol";
+import { Proxy } from "@eth-optimism/contracts-bedrock/contracts/universal/Proxy.sol";
+import { ProxyUpdater } from "./ProxyUpdater.sol";
+import { IProxyAdapter } from "./IProxyAdapter.sol";
+
+/**
+ * @title ProxyAdmin
+ * @notice The ProxyAdmin is a contract associated with each ChugSplashManager that owns the various
+ *         proxies for a given project. The ProxyAdmin delegatecalls into various ProxyAdapter
+ *         contracts that correspond to different proxy types. Using this pattern, the ProxyAdmin
+ *         can universally handle all different proxy types as long as the ProxyAdmin is considered
+ *         the owning address of each proxy.
+ */
 contract ProxyAdmin {
-    function getProxyImplementation(address _proxy, bytes32 _proxyType)
-        public
-        returns (address implementation)
-    {}
+    /**
+     * @notice The storage slot that holds the address of the implementation contract for default
+     *         proxies.
+     *         bytes32(uint256(keccak256('eip1967.proxy.implementation')) - 1)
+     */
+    bytes32 internal constant EIP1967_IMPLEMENTATION_KEY =
+        0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc;
 
-    function upgrade(
-        address _proxy,
-        bytes32 _proxyType,
-        address _implementation
-    ) public {}
+    /**
+     * @notice Address of the ChugSplashRegistry.
+     */
+    ChugSplashRegistry public immutable registry;
+
+    /**
+     * @notice Address of the ProxyUpdater.
+     */
+    address public immutable proxyUpdater;
 
+    /**
+     * @param _registry     Address of the ChugSplashRegistry.
+     * @param _proxyUpdater Address of the ProxyUpdater.
+     */
+    constructor(ChugSplashRegistry _registry, address _proxyUpdater) {
+        registry = _registry;
+        proxyUpdater = _proxyUpdater;
+    }
+
+    /**
+     * @notice Sets new code for the proxy contract's implementation.
+     *
+     * @param _proxy     Address of the proxy, which can be a default proxy or a non-standard proxy.
+     * @param _proxyType The proxy's type. This is the zero-address for default proxies.
+     * @param _code      Creation bytecode to be deployed.
+     */
     function setProxyCode(
-        address _proxy,
+        address payable _proxy,
         bytes32 _proxyType,
-        bytes memory _data
-    ) public {}
+        bytes memory _code
+    ) public {
+        // If no proxy type has been set, use the default proxy.
+        if (_proxyType == bytes32(0)) {
```

Comment:
Although it's slightly less efficient, I think it would be worth (from a code readability standpoint) to have everything go through the adapter system, including the standard proxy. We get to remove the `if` statement here and have everything under a single code path. Doing this will also mean that we get an example ProxyAdapter implementation for the standard proxy, which is a nice bonus.